File: python/parallel.py
import timeit, time
from mpi4py import MPI
import math
import numpy as np
from util import add_and_multiply, step_one, write_to_file, write
from main import a, b, n

if __name__ == '__main__':
    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    p = comm.Get_size()
    step, dim1, s1 = 0, 0, 0
    p_sqrt = int(math.sqrt(p-1))
    p_sqrt_2 = p_sqrt ** 2
    range_per_row = int((p - 1) / p_sqrt)
    block_dim = int(n / p_sqrt)
    dim2 = block_dim
    s2 = range_per_row
    rows = {}
    c = [[0 for i in range(n)] for j in range(n)]

    if rank == 0:
        dest = 0
        write("parallel.txt","matrices a, b", a, b)
        start_time = time.time()
        a, b = step_one(a, b, n)
        for i in range(n):
            a_block, b_block, data = [], [], []
            for j in range(dim1, dim2):
                a_block.append(a[j][step:step + block_dim])
                b_block.append(b[j][step:step + block_dim])
            if len(a_block[block_dim-1]) == block_dim:
                print(a_block, " a BLOCK")
                c_block = [[0 for i in range(block_dim)] for j in range(block_dim)]
                data = [a_block, b_block, c_block]
                dest += 1
                if dest == p:
                    dest = 1
                comm.send(data, dest=dest, tag=1)
            step = step + block_dim
            if (i + 1) % block_dim == 0:
                step = 0
                dim1 += block_dim
                dim2 += block_dim

        c_blocks = []
        d = 0
        for i in range(p-1):
            d = i + 1
            mtx = comm.recv(source=d, tag=d)
            c_blocks.append(mtx)

        for l in range(range_per_row):
            rows[l] = []
            for m in range(s1, s2):
                rows[l].append(c_blocks[m])
            s1 += range_per_row
            s2 += range_per_row

        end_time = time.time()
        print(np.bmat([rows[i] for i in rows.keys()]))
        print("Process finished in ", end_time - start_time)
        write("parallel.txt","result: ",np.bmat([rows[i] for i in rows.keys()]), '-------------------------')
    else:
        data = comm.recv(source=0, tag=1)
        for t in range(n):
            add_and_multiply(data[0], data[1], data[2], block_dim)
            write_to_file('parallel.txt',t+rank, data[0], data[1], data[2])
            left_shift_dest = rank-1+p_sqrt if (rank-1) % p_sqrt == 0 else rank - 1
            left_shift_source = rank + 1 - p_sqrt if rank % p_sqrt == 0 else rank + 1
            comm.send([i[0] for i in data[0]], dest=left_shift_dest, tag=left_shift_dest)
            new_col = comm.recv(source=left_shift_source, tag=rank)
            for i in range(block_dim):
                data[0][i] = data[0][i][1:] + [new_col[i]]

            up_shift_dest = rank - p_sqrt if rank - p_sqrt > 0 else rank + range_per_row * (range_per_row - 1)
            up_shift_source = rank - range_per_row * (range_per_row - 1) if (rank + p_sqrt) > p_sqrt_2 else rank + p_sqrt
            comm.send(data[1][0], dest=up_shift_dest, tag=up_shift_dest)
            new_row = comm.recv(source=up_shift_source, tag=rank)
            data[1] = data[1][1:] + [new_row]

        req = comm.send(data[2], dest=0, tag=rank)

File: go/parallel/parallel.go
package parallel

import (
	"fmt"
	"log"
	"math"
	"os"
	"time"

	"github.com/MilicaPoparic/ntp/go/util"
)

func Parallel(a [][]int, b [][]int, size int, p int) {
	pSqrt := int(math.Sqrt(float64(p)))
	blockDim := size / pSqrt
	// util.WriteMatrix("parallel.txt", "Matrices A, B: ", a, b)

	startTime := time.Now()

	allChans := make([][]chan []int, p)
	for i := range allChans {
		allChans[i] = make([]chan []int, 2) // row, col
		for j := range allChans[i] {
			allChans[i][j] = make(chan []int, blockDim) //row, col dim blockDim
		}
	}

	a, b = util.StepOne(a, b, size)
	dest := 0
	cBlockHolder := make(chan util.CBlockStruct)

	aBlock := make([][]int, blockDim)
	bBlock := make([][]int, blockDim)

	for i := 0; i < size; i += blockDim {
		mtxA := a[i : i+blockDim]
		mtxB := b[i : i+blockDim]
		for j := 0; j < size; j += blockDim {
			for k := range mtxA {
				aBlock[k] = mtxA[k][j : j+blockDim]
				bBlock[k] = mtxB[k][j : j+blockDim]
				if len(aBlock[blockDim-1]) == blockDim {
					// intialize c blok with all zeros
					cBlock := make([][]int, blockDim)
					for i := range cBlock {
						cBlock[i] = make([]int, blockDim)
					}

					data := make([][][]int, 3)
					data[0] = aBlock
					data[1] = bBlock
					data[2] = cBlock

					dest += 1

					aBlock = make([][]int, blockDim)
					bBlock = make([][]int, blockDim)

					leftShiftSource := util.LShiftSource(dest+1, pSqrt)
					upShiftSource := util.UShiftSource(dest, pSqrt, p)

					sendChans := make([]chan []int, 4)
					sendChans[0] = allChans[dest-1][0]            //dest left shift
					sendChans[1] = allChans[dest-1][1]            // dest up shift
					sendChans[2] = allChans[leftShiftSource-1][0] // src left shift
					sendChans[3] = allChans[upShiftSource-1][1]   // src up shift
					go RoutineJob(data, sendChans, size, blockDim, cBlockHolder, dest-1)
				}
			}
		}
	}
	c := make([][]int, size)
	for i := range c {
		c[i] = make([]int, size)
		for j := range c[i] {
			c[i][j] = 0
		}
	}

	for t := 0; t < p; t++ {
		rp := <-cBlockHolder
		for k := 0; k < blockDim; k++ {
			i := rp.Num/pSqrt*blockDim + k
			j := rp.Num % pSqrt * blockDim
			c[i] = append(append(c[i][:j], rp.Matrix[k]...), c[i][j+blockDim:]...)
		}
	}
	for _, row := range c {
		fmt.Println(row)
	}

	elapsed := time.Since(startTime)
	fmt.Println("Process finished in: ", elapsed)
	f, err := os.OpenFile("resources/parallelWeak25.txt",
		os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		log.Println(err)
	}
	f.WriteString(fmt.Sprint(elapsed.Seconds()) + "\n")
	defer f.Close()
}

func RoutineJob(data [][][]int, chans []chan []int, size int, blockDim int, cBlockHolder chan util.CBlockStruct, source int) {
	for t := 0; t < size; t++ {
		util.AddAndMultiply(data[0], data[1], data[2], blockDim)
		if t == size-1 {
			// util.WriteToFile("parallel.txt", source, data[0], data[1], data[2])
			cBlockHolder <- util.CBlockStruct{source, data[2]}
			break
		}
		d1 := make([]int, blockDim)
		for i := 0; i < blockDim; i++ {
			d1[i] = data[0][i][0]
		}

		chans[0] <- d1
		chans[1] <- data[1][0]
		s1 := <-chans[2]
		s2 := <-chans[3]

		for i := 0; i < blockDim; i++ {
			data[0][i] = append(data[0][i][1:], s1[i])
		}
		data[1] = append(data[1][1:], s2)
	}

}
